front/pages/api/v1/w/[wId]/apps/[aId]/runs/index.ts

```diff
@@ -77,6 +82,96 @@ const poll = async ({
   return new Promise(executePoll);
 };
 
+async function pollForRunCompletion(
+  coreAPI: CoreAPI,
+  app: AppType,
+  runId: string
+) {
+  try {
+    await poll({
+      fn: async () => {
+        const run = await coreAPI.getRunStatus({
+          projectId: app.dustAPIProjectId,
+          runId,
+        });
+        if (run.isErr()) {
+          return { status: "error" };
+        }
+        const r = run.value.run;
+        return { status: r.status.run };
+      },
+      validate: (r) => {
+        if (r && r.status == "running") {
+          return false;
+        }
+        return true;
+      },
+      interval: 128,
+      increment: 32,
+      maxInterval: 1024,
+      maxAttempts: 64,
+    });
+  } catch (e) {
+    throw Error(
+      `There was an error polling the run status: runId=${runId} error=${e}`
+    );
+  }
+
+  // Finally refresh the run object.
+  const runRes = await coreAPI.getRun({
+    projectId: app.dustAPIProjectId,
+    runId,
+  });
+  if (runRes.isErr()) {
+    throw Error(`There was an error retrieving the run while polling.`, {
+      cause: runRes.error,
+    });
+  }
+  const run = runRes.value.run;
+  run.specification_hash = run.app_hash;
+  delete run.app_hash;
+  return run;
+}
+
+function extractUsageFromExecutions(
+  block: any,
```

Comment:
I think we generally want to use unknown here and then have an explicit override in the body to with a `as ...`. We do force the type system but in a more precise way

Reply:
I replaced these any (block and executions) by the expected type shape

Reply:
Yeah I do believe we have types for them